File: wargame_platform/apps/wargame_core/lib/wargame_core/hex/coord.ex
defmodule WargameCore.Hex.Coord do
  @moduledoc """
  Axial coordinate system for hex grids.

  Uses axial coordinates (q, r) with computed cube coordinate s = -q - r.
  Based on Red Blob Games implementation guide.

  ## Coordinate System

  Pointy-top hexes with axial coordinates:
  - q: column (increases east)
  - r: row (increases southeast)
  - s: computed as -q - r (for cube coordinate operations)

  ## Direction Indices (pointy-top, counter-clockwise from east)

      0: East      (+1,  0)
      1: Northeast (+1, -1)
      2: Northwest ( 0, -1)
      3: West      (-1,  0)
      4: Southwest (-1, +1)
      5: Southeast ( 0, +1)
  """

  @type t :: %__MODULE__{q: integer(), r: integer()}

  @enforce_keys [:q, :r]
  defstruct [:q, :r]

  @direction_vectors [
    {1, 0},   # 0: East
    {1, -1},  # 1: Northeast
    {0, -1},  # 2: Northwest
    {-1, 0},  # 3: West
    {-1, 1},  # 4: Southwest
    {0, 1}    # 5: Southeast
  ]

  @doc """
  Creates a new hex coordinate.

  ## Examples

      iex> WargameCore.Hex.Coord.new(3, 4)
      %WargameCore.Hex.Coord{q: 3, r: 4}
  """
  @spec new(integer(), integer()) :: t()
  def new(q, r) when is_integer(q) and is_integer(r) do
    %__MODULE__{q: q, r: r}
  end

  @doc """
  Creates a hex coordinate from cube coordinates.
  Validates that q + r + s = 0.

  ## Examples

      iex> WargameCore.Hex.Coord.from_cube(1, -2, 1)
      {:ok, %WargameCore.Hex.Coord{q: 1, r: -2}}

      iex> WargameCore.Hex.Coord.from_cube(1, 1, 1)
      {:error, :invalid_cube_coordinates}
  """
  @spec from_cube(integer(), integer(), integer()) :: {:ok, t()} | {:error, :invalid_cube_coordinates}
  def from_cube(q, r, s) when is_integer(q) and is_integer(r) and is_integer(s) do
    if q + r + s == 0 do
      {:ok, %__MODULE__{q: q, r: r}}
    else
      {:error, :invalid_cube_coordinates}
    end
  end

  @doc """
  Computes the cube s coordinate from axial coordinates.

  ## Examples

      iex> coord = WargameCore.Hex.Coord.new(3, -1)
      iex> WargameCore.Hex.Coord.s(coord)
      -2
  """
  @spec s(t()) :: integer()
  def s(%__MODULE__{q: q, r: r}), do: -q - r

  @doc """
  Calculates the distance between two hex coordinates.
  Distance is the number of steps needed to move from one hex to another.

  ## Examples

      iex> a = WargameCore.Hex.Coord.new(0, 0)
      iex> b = WargameCore.Hex.Coord.new(3, -1)
      iex> WargameCore.Hex.Coord.distance(a, b)
      3
  """
  @spec distance(t(), t()) :: non_neg_integer()
  def distance(%__MODULE__{} = a, %__MODULE__{} = b) do
    max(
      abs(a.q - b.q),
      max(
        abs(a.r - b.r),
        abs(s(a) - s(b))
      )
    )
  end

  @doc """
  Adds two coordinates together.

  ## Examples

      iex> a = WargameCore.Hex.Coord.new(1, 2)
      iex> b = WargameCore.Hex.Coord.new(3, -1)
      iex> WargameCore.Hex.Coord.add(a, b)
      %WargameCore.Hex.Coord{q: 4, r: 1}
  """
  @spec add(t(), t()) :: t()
  def add(%__MODULE__{q: q1, r: r1}, %__MODULE__{q: q2, r: r2}) do
    %__MODULE__{q: q1 + q2, r: r1 + r2}
  end

  @doc """
  Subtracts the second coordinate from the first.

  ## Examples

      iex> a = WargameCore.Hex.Coord.new(4, 1)
      iex> b = WargameCore.Hex.Coord.new(1, 2)
      iex> WargameCore.Hex.Coord.subtract(a, b)
      %WargameCore.Hex.Coord{q: 3, r: -1}
  """
  @spec subtract(t(), t()) :: t()
  def subtract(%__MODULE__{q: q1, r: r1}, %__MODULE__{q: q2, r: r2}) do
    %__MODULE__{q: q1 - q2, r: r1 - r2}
  end

  @doc """
  Scales a coordinate by a factor.

  ## Examples

      iex> coord = WargameCore.Hex.Coord.new(2, 3)
      iex> WargameCore.Hex.Coord.scale(coord, 2)
      %WargameCore.Hex.Coord{q: 4, r: 6}
  """
  @spec scale(t(), integer()) :: t()
  def scale(%__MODULE__{q: q, r: r}, factor) when is_integer(factor) do
    %__MODULE__{q: q * factor, r: r * factor}
  end

  @doc """
  Gets the neighbor in a given direction (0-5).

  ## Direction Indices
  - 0: East
  - 1: Northeast
  - 2: Northwest
  - 3: West
  - 4: Southwest
  - 5: Southeast

  ## Examples

      iex> coord = WargameCore.Hex.Coord.new(0, 0)
      iex> WargameCore.Hex.Coord.neighbor(coord, 0)
      %WargameCore.Hex.Coord{q: 1, r: 0}
  """
  @spec neighbor(t(), 0..5) :: t()
  def neighbor(%__MODULE__{q: q, r: r}, direction) when direction in 0..5 do
    {dq, dr} = Enum.at(@direction_vectors, direction)
    %__MODULE__{q: q + dq, r: r + dr}
  end

  @doc """
  Gets all six neighbors of a hex.

  ## Examples

      iex> coord = WargameCore.Hex.Coord.new(0, 0)
      iex> neighbors = WargameCore.Hex.Coord.neighbors(coord)
      iex> length(neighbors)
      6
  """
  @spec neighbors(t()) :: [t()]
  def neighbors(%__MODULE__{} = coord) do
    Enum.map(0..5, &neighbor(coord, &1))
  end

  @doc """
  Returns the direction vector for a given direction index.

  ## Examples

      iex> WargameCore.Hex.Coord.direction_vector(0)
      {1, 0}
  """
  @spec direction_vector(0..5) :: {integer(), integer()}
  def direction_vector(direction) when direction in 0..5 do
    Enum.at(@direction_vectors, direction)
  end

  @doc """
  Converts the coordinate to a string key for use in maps.

  ## Examples

      iex> coord = WargameCore.Hex.Coord.new(3, -2)
      iex> WargameCore.Hex.Coord.to_key(coord)
      "3,-2"
  """
  @spec to_key(t()) :: String.t()
  def to_key(%__MODULE__{q: q, r: r}) do
    "#{q},#{r}"
  end

  @doc """
  Parses a coordinate from a string key.

  ## Examples

      iex> WargameCore.Hex.Coord.from_key("3,-2")
      {:ok, %WargameCore.Hex.Coord{q: 3, r: -2}}

      iex> WargameCore.Hex.Coord.from_key("invalid")
      {:error, :invalid_key}
  """
  @spec from_key(String.t()) :: {:ok, t()} | {:error, :invalid_key}
  def from_key(key) when is_binary(key) do
    case String.split(key, ",") do
      [q_str, r_str] ->
        with {q, ""} <- Integer.parse(q_str),
             {r, ""} <- Integer.parse(r_str) do
          {:ok, %__MODULE__{q: q, r: r}}
        else
          _ -> {:error, :invalid_key}
        end

      _ ->
        {:error, :invalid_key}
    end
  end

  @doc """
  Converts axial hex coordinate to pixel position (pointy-top orientation).

  ## Parameters
  - coord: The hex coordinate
  - size: Hex size (distance from center to corner in pixels)

  ## Examples

      iex> coord = WargameCore.Hex.Coord.new(1, 1)
      iex> {x, y} = WargameCore.Hex.Coord.to_pixel(coord, 40)
      iex> is_float(x) and is_float(y)
      true
  """
  @spec to_pixel(t(), number()) :: {float(), float()}
  def to_pixel(%__MODULE__{q: q, r: r}, size) when is_number(size) do
    x = size * (:math.sqrt(3) * q + :math.sqrt(3) / 2 * r)
    y = size * (3 / 2 * r)
    {x, y}
  end

  @doc """
  Converts pixel position to the nearest hex coordinate (pointy-top orientation).

  ## Parameters
  - x: Pixel x position
  - y: Pixel y position
  - size: Hex size (distance from center to corner in pixels)

  ## Examples

      iex> coord = WargameCore.Hex.Coord.from_pixel(100.0, 50.0, 40)
      iex> is_struct(coord, WargameCore.Hex.Coord)
      true
  """
  @spec from_pixel(number(), number(), number()) :: t()
  def from_pixel(x, y, size) when is_number(x) and is_number(y) and is_number(size) do
    q = (:math.sqrt(3) / 3 * x - 1 / 3 * y) / size
    r = (2 / 3 * y) / size
    hex_round(q, r)
  end

  @doc """
  Rounds fractional hex coordinates to the nearest hex.

  Uses cube coordinate rounding for accuracy.
  """
  @spec hex_round(number(), number()) :: t()
  def hex_round(q, r) when is_number(q) and is_number(r) do
    s = -q - r

    rq = round(q)
    rr = round(r)
    rs = round(s)

    q_diff = abs(rq - q)
    r_diff = abs(rr - r)
    s_diff = abs(rs - s)

    {final_q, final_r} =
      cond do
        q_diff > r_diff and q_diff > s_diff ->
          {-rr - rs, rr}

        r_diff > s_diff ->
          {rq, -rq - rs}

        true ->
          {rq, rr}
      end

    %__MODULE__{q: final_q, r: final_r}
  end

  @doc """
  Checks if two coordinates are equal.

  ## Examples

      iex> a = WargameCore.Hex.Coord.new(1, 2)
      iex> b = WargameCore.Hex.Coord.new(1, 2)
      iex> WargameCore.Hex.Coord.equal?(a, b)
      true
  """
  @spec equal?(t(), t()) :: boolean()
  def equal?(%__MODULE__{q: q1, r: r1}, %__MODULE__{q: q2, r: r2}) do
    q1 == q2 and r1 == r2
  end

  @doc """
  Returns the opposite direction (180 degrees).

  ## Examples

      iex> WargameCore.Hex.Coord.opposite_direction(0)
      3
      iex> WargameCore.Hex.Coord.opposite_direction(1)
      4
  """
  @spec opposite_direction(0..5) :: 0..5
  def opposite_direction(direction) when direction in 0..5 do
    rem(direction + 3, 6)
  end

  @doc """
  Rotates a direction by a number of steps (each step is 60 degrees).
  Positive steps rotate counter-clockwise.

  ## Examples

      iex> WargameCore.Hex.Coord.rotate_direction(0, 1)
      1
      iex> WargameCore.Hex.Coord.rotate_direction(5, 2)
      1
  """
  @spec rotate_direction(0..5, integer()) :: 0..5
  def rotate_direction(direction, steps) when direction in 0..5 and is_integer(steps) do
    rem(rem(direction + steps, 6) + 6, 6)
  end

  @doc """
  Returns the direction from one coordinate to an adjacent coordinate.

  Returns nil if the coordinates are not adjacent.

  ## Examples

      iex> from = WargameCore.Hex.Coord.new(0, 0)
      iex> to = WargameCore.Hex.Coord.new(1, 0)
      iex> WargameCore.Hex.Coord.direction_to(from, to)
      0

      iex> from = WargameCore.Hex.Coord.new(0, 0)
      iex> to = WargameCore.Hex.Coord.new(2, 0)
      iex> WargameCore.Hex.Coord.direction_to(from, to)
      nil
  """
  @spec direction_to(t(), t()) :: 0..5 | nil
  def direction_to(%__MODULE__{q: q1, r: r1}, %__MODULE__{q: q2, r: r2}) do
    diff = {q2 - q1, r2 - r1}
    Enum.find_index(@direction_vectors, &(&1 == diff))
  end
end

File: wargame_platform/apps/wargame_web/assets/ts/hex/coord.ts
/**
 * Axial coordinate system for hex grids.
 * Based on Red Blob Games implementation guide.
 * @see https://www.redblobgames.com/grids/hexagons/
 */

export interface AxialCoord {
  readonly q: number;
  readonly r: number;
}

/**
 * Immutable hex coordinate using axial coordinate system.
 * Stores q and r; computes s = -q - r when needed.
 */
export class HexCoord implements AxialCoord {
  constructor(
    public readonly q: number,
    public readonly r: number
  ) {}

  /** Cube coordinate s (computed from q and r) */
  get s(): number {
    return -this.q - this.r;
  }

  /** Create a new HexCoord from axial coordinates */
  static from(q: number, r: number): HexCoord {
    return new HexCoord(q, r);
  }

  /** Create a new HexCoord from cube coordinates (validates s = -q - r) */
  static fromCube(q: number, r: number, s: number): HexCoord {
    if (Math.round(q + r + s) !== 0) {
      throw new Error(`Invalid cube coordinates: q + r + s must equal 0`);
    }
    return new HexCoord(q, r);
  }

  /** Check equality with another coordinate */
  equals(other: AxialCoord): boolean {
    return this.q === other.q && this.r === other.r;
  }

  /** Create a string key for use in Maps/Sets */
  toKey(): string {
    return `${this.q},${this.r}`;
  }

  /** Parse a key back to HexCoord */
  static fromKey(key: string): HexCoord {
    const [q, r] = key.split(",").map(Number);
    return new HexCoord(q, r);
  }

  /** Calculate distance to another hex (number of steps) */
  distanceTo(other: AxialCoord): number {
    return HexCoord.distance(this, other);
  }

  /** Static distance calculation between two coordinates */
  static distance(a: AxialCoord, b: AxialCoord): number {
    const aS = -a.q - a.r;
    const bS = -b.q - b.r;
    return Math.max(
      Math.abs(a.q - b.q),
      Math.abs(a.r - b.r),
      Math.abs(aS - bS)
    );
  }

  /** Add two coordinates */
  add(other: AxialCoord): HexCoord {
    return new HexCoord(this.q + other.q, this.r + other.r);
  }

  /** Subtract another coordinate */
  subtract(other: AxialCoord): HexCoord {
    return new HexCoord(this.q - other.q, this.r - other.r);
  }

  /** Scale coordinate by a factor */
  scale(factor: number): HexCoord {
    return new HexCoord(this.q * factor, this.r * factor);
  }

  toString(): string {
    return `HexCoord(${this.q}, ${this.r})`;
  }
}

/**
 * Direction vectors for pointy-top hexes (0 = East, going counter-clockwise)
 */
export const HEX_DIRECTIONS: readonly AxialCoord[] = [
  { q: 1, r: 0 },   // 0: East
  { q: 1, r: -1 },  // 1: Northeast
  { q: 0, r: -1 },  // 2: Northwest
  { q: -1, r: 0 },  // 3: West
  { q: -1, r: 1 },  // 4: Southwest
  { q: 0, r: 1 },   // 5: Southeast
] as const;

/**
 * Get the neighbor of a hex in a given direction
 * @param coord The starting coordinate
 * @param direction Direction index 0-5 (0 = East, counter-clockwise)
 */
export function hexNeighbor(coord: AxialCoord, direction: number): HexCoord {
  const dir = HEX_DIRECTIONS[((direction % 6) + 6) % 6];
  return new HexCoord(coord.q + dir.q, coord.r + dir.r);
}

/**
 * Get all six neighbors of a hex
 */
export function hexNeighbors(coord: AxialCoord): HexCoord[] {
  return HEX_DIRECTIONS.map((dir) => new HexCoord(coord.q + dir.q, coord.r + dir.r));
}

/**
 * Convert axial hex coordinate to pixel position (pointy-top orientation)
 * This is for true axial coordinates where moving in +r direction goes diagonally.
 * @param coord Hex coordinate
 * @param size Hex size (distance from center to corner)
 */
export function hexToPixelAxial(coord: AxialCoord, size: number): { x: number; y: number } {
  const x = size * (Math.sqrt(3) * coord.q + (Math.sqrt(3) / 2) * coord.r);
  const y = size * ((3 / 2) * coord.r);
  return { x, y };
}

/**
 * Convert offset hex coordinate to pixel position (pointy-top, odd-r layout)
 * This creates a rectangular grid where odd rows are shifted right.
 * Used when q=column, r=row in a rectangular map.
 * @param coord Offset coordinate (q=column, r=row)
 * @param size Hex size (distance from center to corner)
 */
export function hexToPixel(coord: AxialCoord, size: number): { x: number; y: number } {
  const hexWidth = Math.sqrt(3) * size;
  const hexHeight = 2 * size;
  const vertSpacing = hexHeight * 0.75;

  // Odd rows shift right by half a hex width
  const xOffset = (coord.r % 2 === 1) ? hexWidth / 2 : 0;

  const x = coord.q * hexWidth + xOffset;
  const y = coord.r * vertSpacing;

  return { x, y };
}

/**
 * Convert pixel position to axial hex coordinate (pointy-top orientation)
 * This is for true axial coordinates.
 * @param x Pixel x
 * @param y Pixel y
 * @param size Hex size
 */
export function pixelToHexAxial(x: number, y: number, size: number): HexCoord {
  const q = ((Math.sqrt(3) / 3) * x - (1 / 3) * y) / size;
  const r = ((2 / 3) * y) / size;
  return hexRound(q, r);
}

/**
 * Convert pixel position to offset hex coordinate (pointy-top, odd-r layout)
 * Used for rectangular maps where q=column, r=row.
 * @param x Pixel x
 * @param y Pixel y
 * @param size Hex size
 */
export function pixelToHex(x: number, y: number, size: number): HexCoord {
  const hexWidth = Math.sqrt(3) * size;
  const hexHeight = 2 * size;
  const vertSpacing = hexHeight * 0.75;

  // First determine the row
  const row = Math.round(y / vertSpacing);

  // Adjust x for the row offset
  const xOffset = (row % 2 === 1) ? hexWidth / 2 : 0;
  const col = Math.round((x - xOffset) / hexWidth);

  // Do a more precise check by testing this hex and neighbors
  // to find which hex center is actually closest
  const candidates = [
    { q: col, r: row },
    { q: col - 1, r: row },
    { q: col + 1, r: row },
    { q: col, r: row - 1 },
    { q: col, r: row + 1 },
    { q: col - 1, r: row - 1 },
    { q: col + 1, r: row - 1 },
    { q: col - 1, r: row + 1 },
    { q: col + 1, r: row + 1 },
  ];

  let closest = candidates[0];
  let closestDist = Infinity;

  for (const candidate of candidates) {
    const center = hexToPixel(candidate, size);
    const dist = Math.sqrt((x - center.x) ** 2 + (y - center.y) ** 2);
    if (dist < closestDist) {
      closestDist = dist;
      closest = candidate;
    }
  }

  return new HexCoord(closest.q, closest.r);
}

/**
 * Round fractional hex coordinates to nearest hex
 */
export function hexRound(q: number, r: number): HexCoord {
  const s = -q - r;

  let rQ = Math.round(q);
  let rR = Math.round(r);
  let rS = Math.round(s);

  const qDiff = Math.abs(rQ - q);
  const rDiff = Math.abs(rR - r);
  const sDiff = Math.abs(rS - s);

  if (qDiff > rDiff && qDiff > sDiff) {
    rQ = -rR - rS;
  } else if (rDiff > sDiff) {
    rR = -rQ - rS;
  }

  return new HexCoord(rQ, rR);
}

/**
 * Get all hexes within a given range of a center hex
 */
export function hexesInRange(center: AxialCoord, range: number): HexCoord[] {
  const results: HexCoord[] = [];
  for (let q = -range; q <= range; q++) {
    for (let r = Math.max(-range, -q - range); r <= Math.min(range, -q + range); r++) {
      results.push(new HexCoord(center.q + q, center.r + r));
    }
  }
  return results;
}

/**
 * Get hexes forming a ring at a given radius from center
 */
export function hexRing(center: AxialCoord, radius: number): HexCoord[] {
  if (radius === 0) {
    return [new HexCoord(center.q, center.r)];
  }

  const results: HexCoord[] = [];
  let hex = new HexCoord(
    center.q + HEX_DIRECTIONS[4].q * radius,
    center.r + HEX_DIRECTIONS[4].r * radius
  );

  for (let i = 0; i < 6; i++) {
    for (let j = 0; j < radius; j++) {
      results.push(hex);
      hex = hexNeighbor(hex, i);
    }
  }

  return results;
}

/**
 * Draw a line between two hexes using linear interpolation
 */
export function hexLineDraw(a: AxialCoord, b: AxialCoord): HexCoord[] {
  const n = HexCoord.distance(a, b);
  if (n === 0) {
    return [new HexCoord(a.q, a.r)];
  }

  const results: HexCoord[] = [];
  const aS = -a.q - a.r;
  const bS = -b.q - b.r;

  for (let i = 0; i <= n; i++) {
    const t = i / n;
    const q = a.q + (b.q - a.q) * t;
    const r = a.r + (b.r - a.r) * t;
    results.push(hexRound(q, r));
  }

  return results;
}
